dali/pipeline/operators/arithmetic/expression_tree.h

```diff
@@ -0,0 +1,240 @@
+  }
+}
+
+/**
+ * @brief base class for nodes of ExpressionTree
+ *
+ * TODO(klecki): Using unique_ptrs for ownership, we may generalize it to DAG, that would require
+ * some changes
+ */
+class ExprNode {
+ public:
+  virtual const std::string &GetOp() const {
+    DALI_FAIL("No op in this expression node.");
+  }
+
+  virtual std::string GetNodeDesc() const {
+    return GetOutputDesc();
+  }
+
+  virtual std::string GetOutputDesc() const {
+    auto op_type = TypeTable::GetTypeInfo(GetTypeId()).name();
+    std::string result = GetAbbreviation(GetNodeType());
+    result += IsScalarLike(GetShape()) ? "C:" : "T:";
+    return result + op_type;
+  }
+
+  virtual NodeType GetNodeType() const = 0;
+
+  std::unique_ptr<ExpressionImplBase> impl;
+
+  void SetTypeId(DALIDataType type_id) {
+    type_id_ = type_id;
+  }
+
+  DALIDataType GetTypeId() const {
+    return type_id_;
+  }
+
+  void SetShape(const kernels::TensorListShape<> &shape) {
+    shape_ = shape;
+  }
+
+  const kernels::TensorListShape<> &GetShape() const {
+    return shape_;
+  }
+
+  virtual int GetSubexpressionCount() const {
+    return 0;
+  }
+
+  virtual ExprNode &operator[](int i) {
+    DALI_FAIL("No subexpressions for this node.");
+  }
+
+  virtual const ExprNode &operator[](int i) const {
+    DALI_FAIL("No subexpressions for this node.");
+  }
+
+  virtual ~ExprNode() = default;
+
+ private:
+  DALIDataType type_id_ = DALI_NO_TYPE;
+  kernels::TensorListShape<> shape_ = {};
+};
+
+class ExprFunc : public ExprNode {
+ public:
+  explicit ExprFunc(const std::string &op) : op_(op) {}
+
+  ExprFunc(const std::string &op, int num_input) : op_(op) {
+    subexpr_.resize(num_input);
+  }
+
+  const std::string &GetOp() const override {
+    return op_;
+  }
+
+  std::string GetNodeDesc() const override {
+    auto op_type = TypeTable::GetTypeInfo(GetTypeId()).name();
+    std::string result = op_ + (IsScalarLike(GetShape()) ? ":C:" : ":T:") + op_type + "(";
+    for (int i = 0; i < GetSubexpressionCount(); i++) {
+      result += (*this)[i].GetOutputDesc();
+      if (i < GetSubexpressionCount() - 1) {
+        result += " ";
+      }
+    }
+    result += ")";
+    return result;
+  }
+
+  NodeType GetNodeType() const override {
+    return NodeType::Function;
+  }
+
+  int GetSubexpressionCount() const override {
+    return subexpr_.size();
+  }
+
+  void AddSubexpression(std::unique_ptr<ExprNode> expr) {
+    subexpr_.push_back(std::move(expr));
+  }
+
+  ExprNode &operator[](int i) override {
+    return *subexpr_[i];
+  }
+
+  const ExprNode &operator[](int i) const override {
+    return *subexpr_[i];
+  }
+
+ private:
+  std::string op_;
+  std::vector<std::unique_ptr<ExprNode>> subexpr_;
+};
+
+class ExprTensor : public ExprNode {
+ public:
+  explicit ExprTensor(int mapped_input) : mapped_input_(mapped_input) {}
+
+  NodeType GetNodeType() const override {
+    return NodeType::Tensor;
+  }
+
+  int GetMappedInput() const {
```

Comment:
Shouldn't it be `GetInputIndex()`? The "mapped" part doesn't really add much to the clarity.

Reply:
Done